ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java

```diff
@@ -148,6 +169,104 @@ public FileOutput(ValueProvider<String> outputPrefix, OutputFileFormat format,
     }
   }
 
+  /**
+   * Implementation of writing to local or remote files.
+   *
+   * <p>For details of the intended behavior for file paths, see:
+   * https://github.com/mozilla/gcp-ingestion/tree/master/ingestion-beam#output-path-specification
+   */
+  public static class AvroOutput extends Write {
+
+    private final ValueProvider<String> outputPrefix;
+    private final Duration windowDuration;
+    private final ValueProvider<Integer> numShards;
+    private final Compression compression;
+    private final InputType inputType;
+    private final AvroSchemaStore schemaStore;
+    private final PubsubMessageRecordFormatter formatter = new PubsubMessageRecordFormatter();
+    private final GenericRecordBinaryEncoder binaryEncoder = new GenericRecordBinaryEncoder();
+    private final BinaryRecordFormatter binaryFormatter = new BinaryRecordFormatter();
+
+    /** Public constructor. */
+    public AvroOutput(ValueProvider<String> outputPrefix, Duration windowDuration,
+        ValueProvider<Integer> numShards, Compression compression, InputType inputType,
+        ValueProvider<String> schemasLocation) {
+      this.outputPrefix = outputPrefix;
+      this.windowDuration = windowDuration;
+      this.numShards = numShards;
+      this.compression = compression;
+      this.inputType = inputType;
+      this.schemaStore = AvroSchemaStore.of(schemasLocation);
+    }
+
+    @Override
+    public WithErrors.Result<PDone> expand(PCollection<PubsubMessage> input) {
+      final PCollectionView<AvroSchemaStore> schemaSideInput = input.getPipeline()
+          .apply(Create.of(schemaStore)).apply(View.asSingleton());
+
+      final TupleTag<PubsubMessage> successTag = new TupleTag<PubsubMessage>() {
+      };
+      final TupleTag<PubsubMessage> errorTag = new TupleTag<PubsubMessage>() {
+      };
+
+      ValueProvider<DynamicPathTemplate> pathTemplate = NestedValueProvider.of(outputPrefix,
+          DynamicPathTemplate::new);
+      ValueProvider<String> staticPrefix = NestedValueProvider.of(pathTemplate,
+          value -> value.staticPrefix);
+
+      ParDo.MultiOutput<PubsubMessage, PubsubMessage> intoGenericRecord = ParDo
```

Comment:
This should be renamed to something like `encodeAvro`

Reply:
I'd like to see this factored out to a PTransform that extends MapElementsWithErrors. That will hide most of the boilerplate for handling the exception (unless there's additional nuance here I missed).

Reply:
I made a note in https://github.com/mozilla/gcp-ingestion/pull/470#issuecomment-478786548 about this ParDo, since I originally tried to write it as a PTransform. I had difficulty including side-input into the PTransform, which is why I use the ParDo here.

I could probably refactor this so the PTransform contains a static schema store that it uses independently from the one in this scope. Would that be preferable?

Reply:
I don't feel strongly about it. The side input detail seems like sufficient motivation to define it directly as a ParDo here. I'd be happy if you just add a comment indicating that we're not using  MapElementsWithErrors on account of the side input.